### ISSUE TITLE
Potential crossover with animatplot?

### ISSUE THREAD

**Issue by TomNicholas** [jbusecke/xmovie#2]
Hi,

This is really cool, and I've been trying to make something fairly similar using [animatplot](https://github.com/t-makaro/animatplot) and integrating it [directly into xarray](https://github.com/pydata/xarray/pull/2729). I'm just wondering how much crossover there might be here? We're both trying to animate xarray objects, use xarray's plotting methods, output as gifs (you're also doing movies), and have flexibility to use libraries like cartopy in conjunction.

On the other hand you're essentially stitching frames together from outside xarray, whereas I'm trying to write animation methods into xarray itself.

What do you think?

**Comment by jbusecke** [comment]
Hi Tom, 

Thanks for checking out my code. I think you are right. There is quite a lot of overlap. Your code seems to be more mature and really my package is just a thin wrapper to have the simplest (its not the quickest) start into movies. 
The main features of xmovie really are the presets (which I plan to extend over time) and sizing the frames right in terms of pixel size. 
I am not planning to add anything like interactive plotting. I really want to emphasize the simplest way to get quick movies out of an xarray or numpy array. Do you plan to add the ability to render out frames and produce high quality movies? Then I think we should certainly talk about an effort to merge.

I think your planned integration within xarray would be phenomenal! 

For now I will make sure to mention animatplot in the readme, so that people are aware of the choices. What do you think?

**Comment by dcherian** [comment]
Another (possible) big difference here is parallel (dask) abilities.

I think @TomNicholas is thinking of

```
movie = da.plot.contourf(..., animate='time')
movie.to_gif()
```

What would be interesting is if we could pull out individual frames from `movie`, render and save to disk in parallel using dask and then combine using `ffmpeg` which seems to be where `xmovie` is going (? I'm looking at commented out code here). This could be an alternative to `to_gif()` then.

**Comment by jbusecke** [comment]
I had that running earlier (nice and fast). In fact that was how xmovie was started. 
I however ran into super weird errors with flickering cbar labels, and was not able to diagnose what was going on.
Hence I pushed this feature back. But I definitely want to bring it back. 
How about Ill put together a PR soon and document the errors, maybe folks out here have an idea what is wrong?

**Comment by jbusecke** [comment]
I have another PR almost ready that will take care of all the ffmpeg stuff with ‘subprocesses’. I would be happy to work on integrating any of that logic to animatplotlib/ xarray aswell.

**Comment by TomNicholas** [comment]
@jbusecke I think that what @dcherian suggested

> What would be interesting is if we could pull out individual frames from movie, render and save to disk in parallel using dask and then combine using ffmpeg which seems to be where xmovie is going (? I'm looking at commented out code here). This could be an alternative to to_gif() then.

Could make sense as a PR to animatplot, or maybe to xarray.animation. animatplot is already storing the information for each frame of the plot in a sensible way, it shouldn't be hard to use whatever implementation you choose here as a new method, e.g. `animatplot.animation.to_movie()`. If you submit it I'll review it.

> The main features of xmovie really are the presets (which I plan to extend over time) and sizing the frames right in terms of pixel size.

The presets I think are the biggest difference. The idea of animatplot integration into xarray is to make animations which look exactly the same as matplotlib figures, just animated. When I get around to trying to animate 2D plots then I will try to make sure it's compatible with cartopy, but the result would still just be an animated cartopy plot, not any extra backgrounds or anything.

**Comment by jbusecke** [comment]
Awesome. I am still working out some kinks in the CI but once the functionality is ready in xmovie we can chat about how/where to implement it. How does that sound?

Julius Busecke

**Comment by TomNicholas** [comment]
Hi @jbusecke and @dcherian, I've had a different idea about how to implement similar functionality to xmovie and animatplot, but in a more general way using jupyter widgets, and I'm wondering what you both think?

It's not hard to use the `Play` widget from ipywidgets to create an updating animation in a jupyter notebook, I've been messing around with general interactivity [over here](https://github.com/pydata/xarray/issues/3709).

![](https://user-images.githubusercontent.com/35968931/72755630-de731680-3bc2-11ea-9d0f-46da96d6efda.gif)

Basically what if this widget also had a record button?

If we could only save this animation, it would a very general way to create animations. Matplotlib is being used here but you can use the `Play()` widget to animate any kind of output, so as long as you can save it you can animate it. You could have multiple plots animated together, you only need to output them into the same jupyter cell. Anything else about the animation that you want to evolve (like making the Earth spin) just has to be parameterized by the same widget. This also isn't at all xarray-specific.

I'm imagining a `Record()` widget (or ipython cell magic command or context manager or whatever is easiest) that:

1. Uses `nbconvert`'s `MarkdownExporter` or similar to save the output of the current cell as a `.png`,
2. Saves each frame as another `.png` by watching the widget to see when it updates,
3. Once the widget does its final update then uses `ffmpeg` to stitch the resulting images together (like xmovie already does).

So ideally the API could just be
```python
import xarray.interactive
from ipyrecord import Record

da.interactive.isel(xplot.plot, time=Record(file='my_movie.gif')) 
```

Can you see any downsides of doing it like this? Any reasons why this wouldn't work?
The only downside I can see is that you can probably couldn't use dask to save the frames in parallel.

**Comment by jbusecke** [comment]
I think the dask argument is a pretty strong one. @dcherian recently proposed a promising way to implement it, I just did not have any time to work on xmovie lately. 

This looks extremely sleek though, and I would love to use it. Xmovie is (supposed to be) designed quite modular, so I wonder if we could carry two mechanisms to actually render images? This version would be for fast printing plots, whereas the dask version could work for movies where each frame tales several secs to render?

**Comment by TomNicholas** [comment]
> I think the dask argument is a pretty strong one.

If you were happy to suppress output to the notebook viewer it might be possible to use dask with this approach too - it's only the sequential output to the notebook that forces the saving to be sequential. It might involve passing a sort of fake widget which simultaneously saves one frame for each possible value.

> This looks extremely sleek though, and I would love to use it.

I'll have a go anyway and see how far I can get :)

**Comment by jbusecke** [comment]
Hey Tom, just wantedt to get back to this issue and see if we can close it? I have convinced myself that this package can occupy a somewhat parallel niche to more flexible/performant/complex tools like hvplot etc. Do you think that recent developments there enable us to close this issue?

**Comment by TomNicholas** [comment]
I think that recent developments in hvplot (see [discussion on xarray here](https://github.com/pydata/xarray/issues/3709#issuecomment-747097516)) render a lot of this discussion moot, and would be the more sensible place for further development, at least of the features I was imagining.
